BUG: Fix scope get to use hashmap lookup instead of list lookup

LeeTZ opened the pull request:
Overview
------
This PR aims to fix a bug in looking up an `op` in `scope`.  The current implementation is using a list lookup to look up items in `iter(_items)`. This is slower than a hashmap lookup, also it may cause bugs for objects if their equality is ill-defined.

Example
-------
```
a = ibis.interval(days=1).op()
b = ibis.interval(hours=1).op()
>>> a
Literal(1, Interval*value_type=int8, unit='D', nullable=True)
>>> a == b
True
```
In this case, `Literal` is ill-defined for equality. And if we have only `a` in scope`. `b in iter(scope._items)` will be `True` while `b in scope._items` return false. 

Tests
-------
This is tested in `ibis/pandas/tests/tes_core.py` with a new test case `test_scope_look_up`.

jreback commented on ibis/expr/scope.py:
```diff
@@ -122,10 +122,10 @@ def merge_scope(self, other_scope: 'Scope', overwrite=False) -> 'Scope':
         result = Scope()
 
-        for op in self.items():
+        for op in self._items:
```
should be `_items.keys()` no?

LeeTZ replied:
I am not sure why using `.keys()` is better here, `_items` is a dict and `op` is its key. Isn't a simple `in` enough?

jreback replied:
because its not obvious that you _items is actually a dict from just looknig here, whereas .keys() is very explict

LeeTZ replied:
OK, fixed it.

jreback commented on ibis/expr/scope.py:
```diff
@@ -122,10 +122,10 @@ def merge_scope(self, other_scope: 'Scope', overwrite=False) -> 'Scope':
             result[op] = self[op]
 
-        for op in other_scope.items():
+        for op in other_scope._items:
```
same

jreback commented on ibis/expr/scope.py:
```diff
@@ -153,7 +153,7 @@ def merge_scopes(
         """
         result = Scope()
-        for op in self.items():
+        for op in self._items:
```
same

icexelloss commented:
I found this class to be a little confusing. It looks like we are emulating dict interface by implementing `__setitem__` and `__getitem__` in addition to `get_value` and `merge_scope`, but:

* It didn't implement other dict interfaces, such as `__contains__`
* Scope.items() returns the keys in the scope (the ops), but dict.items() returns the dict entries (k, v pair), why do we make them different if we are emulating dict interface?
* It's not obviously clear to me if this class is defined to be mutable or immutable - on the one hand, `merge_scope` seems to not mutating exising scope, but `__setitem__` indicates this class is mutable.

I think clarifying these questions would be helpful to make it more clear how to use this class.

LeeTZ commented:
Thanks, @icexelloss ,
This class Scope is refactored from a dict. So it makes sense to implement some interfaces at the first glance. However I have to point out that from the perspective of a user of this class, the key is important, but the value is not. Since value is a data structure that we made up to wrap time context and value, we do need API to get the real value of the cached result, but not necessarily the API to get `{"value":v,"timecontext":t}`.  
Also, for `set`, most scope usages for setting in Ibis is merging scopes generated by different functions. This class tried to preserve this behavior so `merge` is really the `set` method for this class.

So given this `get` and `set` use case, I will say without `__get_item__` , `__set_item__` or any dict like interface, this class works. Dict interfaces are helpful in some implementation of the internal method inside this class, but it does add the ambiguity of how to use this class.

For @icexelloss's question about `__contain__` and `items()`, would you mind sharing your thoughts?@jreback

LeeTZ commented:
I refactored a little bit.
  - removed `__get_item__`. As I commented above, we don't care about the object associated with `op`, it is an implementation detail for this class and we should hide it from users. We exposed `get_value` to get the cached result for a given `op`.
  - removed `__set_item__`. We want to make this class immutable and always use `merge` for adding new items into `Scope`. This is the usage case in Ibis everywhere (by calling `toolz.merge`, before time context refactor )
  - add `__contains__` One thing we need is to test whether an `op` is in Scope. This replaces `op in scope._items.keys()` or `scope.get_value(op)`
  - I think this class doesn't fully emulate a dict interface since we have made it immutable. thus not all dict interfaces are implemented.


To make this class clearer, these are the exposed API for `Scope` class:
1. `new` a 'Scope`: user should use `Scope()` to create an empty `Scope`, or call `make_scope` with a pair of `op, result, time context` to create a new Scope with one item in it.
2. `get` in `Scope`:  user needs to call `get_value` to get the cached result for a given `op`.
3. `set` in `Scope`:  user always uses `merge_scope` or `merge_scopes` to set values.

icexelloss commented on ibis/expr/scope.py:
```diff
@@ -48,14 +48,8 @@ class Scope:
-    def __setitem__(self, op: Node, value: Any) -> None:
-        self._items[op] = value
+    def __contains__(self, op):
+        return True if op in self._items.keys() else False
```
```
def __contains__(self, op):
    return op in self.items
```
?

LeeTZ replied:
Yes

icexelloss commented on ibis/expr/scope.py:
```diff
@@ -122,17 +116,17 @@ def merge_scope(self, other_scope: 'Scope', overwrite=False) -> 'Scope':
 
-        for op in self.items():
-            result[op] = self[op]
+        for op in self._items.keys():
```
Is this basically what you want?
https://stackoverflow.com/questions/6354436/python-dictionary-merge-by-updating-but-not-overwriting-if-value-exists

```
new_items = {}

if overwrite:
    new_items = dict(list(self._items.items()) + list(other._items.items()))
else:
    new_items = dict(list(other._items.items()) + list(self._items.items()))

return new Scope(new_items)
```

LeeTZ replied:
This is nice but to update we cannot simply do these. We should think about time context and the logic is covered in `result.get_value(op, v.timecontext) is None`

icexelloss replied:
I see. Ok what you have is fine.

jreback replied:
why don't we define
```
def __iter__(self):
    return iter(self._items.keys())
```

then 
```
for op in self:
   ...
````

will just work

or is this more confusing?

LeeTZ replied:
This is fine, scope is indeed iterable, I will do that

icexelloss reviewed (changes requested):
Left some comments

icexelloss commented:
I think it makes sense. Emulating dict interface is nice but no necessarily IMO.

icexelloss commented on ibis/expr/scope.py:
```diff
@@ -48,14 +48,8 @@ class Scope:
-    def __setitem__(self, op: Node, value: Any) -> None:
-        self._items[op] = value
+    def __contains__(self, op):
+        return op in self._items.keys()
```
why not use 

```
return op in self._items?
```

icexelloss replied:
Oh I see Jeff's comments. Ok this is fine.

jreback replied:
hmm, no i agree this should be in `self._items`, only *iteration* should use the keys

icexelloss replied:
For my knowledge, is `k in self._items.keys` O(n) or O(1)? The type of self._items.keys is `dict_keys` so it's not clear to me if the `__contains__` method of `dict_key` is linear or constant.

LeeTZ replied:
I looked around and seems that in Python 3.x, `k in self._items` and `k in self._items.keys()` are equivalent and they are all `O(1)`.

LeeTZ replied:
I tested with a Python 3.6 kernel
```
>>> dic = dict.fromkeys(range(10**5))
>>> %timeit 10000 in dic
35.8 ns ± 0.757 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
>>> %timeit 10000 in dic.keys()
74.7 ns ± 0.385 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
seems `keys()` is still 2x slower than testing `in` for dict directly

icexelloss reviewed (approved):
LGTM

icexelloss reviewed (changes requested):
Small comments

LeeTZ commented on ibis/expr/scope.py:
```diff
@@ -116,10 +119,10 @@ def merge_scope(self, other_scope: 'Scope', overwrite=False) -> 'Scope':
             result._items[op] = self._items[op]
 
-        for op in other_scope._items.keys():
+        for op in other_scope._items:
```
It is, thanks

jreback reviewed (changes requested):
can you add this PR number to the original scope update in the release notes

icexelloss reviewed (approved):
+1

jreback reviewed (approved):
lgtm. let's merge on green.

LeeTZ commented:
Thank you all for reviewing! Green now @jreback .
